### PR TITLE
Repair inconsistent nullability nullability on TypeConverter (#63186)

### LINKED ISSUE #63186: Inconsistent nullability change in TypeConverter.CanConvertTo resulting in a breaking API change with .NET 6.0
In .NET 6.0, the nullability of the overload `TypeConverter.CanConvertTo(ITypeDescriptorContext? context, Type destinationType)` was changed. Now, the `destinationType` paramter is also nullable. The same change was not done with `TypeConverter.CanConvertTo(Type destinationType)` and `TypeConverter.ConvertTo`

https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.typeconverter.canconvertto?view=net-6.0
`public virtual bool CanConvertTo (System.ComponentModel.ITypeDescriptorContext? context, Type? destinationType);`

https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.typeconverter.canconvertto?view=net-5.0
`public virtual bool CanConvertTo (System.ComponentModel.ITypeDescriptorContext context, Type destinationType);`

I believe that `destinationType` being nullable is an API bug introduced in .NET 6.0 since it is a) inconsistent with the remaining API, b) not documented as a use case in MSDN, and c) appears to defeat the purpose of the conversion API since you need to know the destination type in order to make an informed choice.

Note that the new API breaks existing code based on .NET 5.0 (custom TypeConverter implementations).

Note that according to git, the entire nullability of TypeConverter was done with ##54961 after .NET 5.0 was already released, so I can't really backtrack this to the root cause.

### PR DESCRIPTION
**Pull request by MichaelKetting**
Fixes #63186

- [ ] new ticket to update the docs (see https://github.com/dotnet/runtime/pull/63874#discussion_r801717043)

- [ ] new ticket to discuss if a change to `TypeConverter.CanConvertFrom(Type sourceType)` to nullable with `NotNullWhen(true)` is also wanted. Note: I've also found one change on src/ImageConverter.CanConvertFrom and ref/ImageFormatConverter.CanConvertFrom and placed them in a dedicated commit. I can create a separate ticket for this if you prefer.

### CODE REVIEW

**Comment by MichaelKetting** [comment]
@maryamariyan I believe my PR is running into #63439 (DllImporter fails on MacOS for Arm64) but is otherwise I'm ready for review. I'll rebase on top of main once #63439 is done and check again that my changes didn't also break something but figured I'll let you know in case there's something else to be done before that.

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs
Comment (eerhardt) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs:41]: I looked into this more, and I think the advice I gave on https://github.com/dotnet/runtime/issues/63186#issuecomment-1004235914 wasn't 100% correct. Now that I look at it again, I think this should be annotated as:

```C#
public virtual bool CanConvertTo(ITypeDescriptorContext? context, [NotNullWhen(true)] Type? destinationType)
```

And then to fix the inconsistency, the above CanConvertTo overload should be annotated the same:

```C#
public bool CanConvertTo([NotNullWhen(true)] Type? destinationType)
```

Note that `ConvertTo` doesn't need to change. It doesn't allow for `null`, and will throw if someone passes `null`.

This way, people can write code like:

```C#
Type? someType = ...;

if (typeConverter.CanConvertTo(someType))
{
    var converted = typeConverter.ConvertTo(someType);
}
```

And they won't get any nullability warnings.

@krwq - since you originally added nullability here - do you have any thoughts?

Comment (MichaelKetting) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs:41]: @eerhardt Thanks for the input. That does make sense. 

Should I expand the original ticket to also cover the `sourceType` parameter on CanConvertFrom?

Comment (eerhardt) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs:41]: > Should I expand the original ticket to also cover the sourceType parameter on CanConvertFrom?

No. Let's open a new issue for CanConvertFrom. The scenario there is a bit different because presumably you've gotten the type from calling `something.GetType()`. ("From" implies you have something already that you are going to convert from.) I'd rather keep these separate, so we can get the CanConvertTo changes in without being held up by any issues with CanConvertFrom.

Comment (MichaelKetting) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs:41]: @eerhardt Sure thing. Just let me know how you want to handle the ImageConverter-fix for CanConvertFrom.

Comment (MichaelKetting) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs:41]: @eerhardt and @krwq Do you already have an update on the final decision on how to proceed with CanConvertTo (see https://github.com/dotnet/runtime/pull/63874#discussion_r786937186 , syntax update with `NotNullWhen` instead of not null)?

Comment (krwq) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs:41]: I guess owner of this library would have better context on the intent of this method. If we don't want to throw on null destinationType then @eerhardt's suggestion (NotNullWhen) is correct and we avoid breaking changes. If we are ok with changing it to non-nullable type then that's fine too - it's a matter of deciding if we think it's likely a (consumer's) product bug to pass null or is it a valid scenario. I.e. if people sometimes pass in something like `obj?.GetType()` then I would go ahead with @eerhardt suggestion. IIRC when I originally annotated I changed this method multiple times as I wasn't sure and neither implementation nor tests nor docs convinced me either way.

Comment (eerhardt) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs:41]: @MichaelKetting - I think we can go forward with my above suggestion. Any plans on moving this forward?

Comment (MichaelKetting) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs:41]: @eerhardt Great, yes, will be updating the PR in the next couple of days / the weekend to go with the NotNullWhen solution.

Since the original null from way back changes did not update the docs, should I take the opportunity and do something there as well by adding parameter documentation and null guides?

Comment (eerhardt) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs:41]: You can, but those docs are contained in a separate repo. Updating the xml doc above doesn't make it into the official docs.

Here are the official docs:
https://github.com/dotnet/dotnet-api-docs/blob/d90566928a7c5f6070f942a89d9be588647a7105/xml/System.ComponentModel/TypeConverter.xml#L349-L407

Comment (MichaelKetting) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs:41]: @eerhardt Ah, okay, there they are :) This would then be a separate ticket and PR, correct, since they're separate repos? Which means I'll do the code first and once that's done and approved, I'll check out the docs.

Comment (eerhardt) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeConverter.cs:41]: Yes, that would be a separate PR

File: src/libraries/System.ComponentModel.TypeConverter/ref/System.ComponentModel.TypeConverter.cs
Comment (eerhardt) [inline src/libraries/System.ComponentModel.TypeConverter/ref/System.ComponentModel.TypeConverter.cs:1294]:
```suggestion
        public bool CanConvertTo([System.Diagnostics.CodeAnalysis.NotNullWhen(true)] System.Type? destinationType) { throw null; }
```

Comment (MichaelKetting) [reply src/libraries/System.ComponentModel.TypeConverter/ref/System.ComponentModel.TypeConverter.cs]: @eerhardt Thanks! Just to let you know, the PR is still in draft-mode, so the review is welcome but may find too many things that aren't done yet.

**Comment by MichaelKetting** [comment]
@eerhardt I'll update the PR from Draft to Ready-for-Review in the evening. Want to update the docs in the PR first. Until then: There's some crashes / failures in helix but I can't pinpoint (yet) if that's my changes or a general instability.

**Comment by MichaelKetting** [comment]
@eerhardt I've rebased on the latest main and should be done. The errors in the pipeline / helix seem to be unrelated to the changes in this PR and I've found the same issues also in at least one other PR (https://github.com/dotnet/runtime/pull/65944/checks?check_run_id=5365597384)

**Comment by eerhardt** [comment]
Failures are #65890.

**Review by eerhardt — APPROVED** [review]
Thanks for the contribution, @MichaelKetting!

**Pull request merged**
